### ISSUE TITLE
[Feature Request] Support type hints

### ISSUE THREAD

**Issue by q-wertz** [rbarrois/python-semanticversion#138]
It would be nice if this package supports type hints (for reference e.g. [PEP 484](https://peps.python.org/pep-0484/) and [PEP 561](https://peps.python.org/pep-0561/)).

Can help you with the type annotation work.

**Comment by rbarrois** [comment]
Indeed, that's a good idea! Although the library currently provides support for old Python version — it's a semantic versioning library after all 😉

I don't have lots of time right now (either for coding or reviewing), but hope to be able to look at this over the next few weeks ;)

**Comment by q-wertz** [comment]
Nice to hear :)

You want to keep the python 2.7 support?

Maybe one could use [PEP 484](https://peps.python.org/pep-0484/#suggested-syntax-for-python-2-7-and-straddling-code) or don't add the type hints in the code but generate a [stub file](https://mypy.readthedocs.io/en/stable/stubs.html).

**Comment by plannigan** [comment]
I started playing around with this. The existence of `partial` make things more complicated, but likely still doable. Since it and a number of other things are already marked for removal, I think it would be best work towards that new major version, then include type hints. The decision to drop unsupported older python versions, as part of the new major version, would have an effect on how the hints we implemented.

**Comment by rbarrois** [comment]
@plannigan That's a great approach indeed, I totally support it. Let me know if you have some code to use as a basis, otherwise I'll follow your recommendation for version 3.0 :)

**Comment by plannigan** [comment]
Actually, I do. Since I didn't get a response, I made an [independent implementation](https://github.com/plannigan/hyper-bump-it/blob/b6c8aaf2bdd9ad8923e2ba739b855a506a573d81/hyper_bump_it/_hyper_bump_it/version.py#L1) in the tool I was using this library in. It isn't a full implementation because I `hyper-bump-it` doesn't need the reversion range specification, but I'm happy with how it ended up.

I can put it up as a PR, but I'm not sure how much work it would be to make the rest of the code base compatible with the API changes.
